AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSTokenExchange.m

```diff
@@ -3,43 +3,129 @@
+                        completion([[MSTokensResponse alloc] initWithTokens:nil], error);
+                    }
+                    
+                    [MSTokenExchange saveToken:[[MSTokenResult alloc] initWithDictionary:jsonDictionary[kMSTokens][0]]];
+                    // Create token result object.
+                    MSTokensResponse *tokens = [[MSTokensResponse alloc] initWithDictionary:jsonDictionary];
+                    completion(tokens, error);
+                }
+            }];
+    }
+    else
+    {
+        NSError *error;
+        completion([[MSTokensResponse alloc] initWithTokens:@[cachedToken]], error);
+    }
+}
+
++ (void)saveToken:(MSTokenResult *)tokenResult {
+    NSString *tokenString = [tokenResult serializeToString];
+    BOOL success = [MSKeychainUtil storeString:tokenString forKey:[MSTokenExchange tokenKeyNameForPartition:tokenResult.partition]];
+    if (success) {
+        MSLogDebug([MSDataStore logTag], @"Saved token in keychain for the partitionKey : %@.", tokenResult.partition);
+    } else {
+        MSLogWarning([MSDataStore logTag], @"Failed to save the token in keychain for the partitionKey : %@.", tokenResult.partition);
+    }
+}
 
-  // Payload.
-  NSError *jsonError;
-  NSData *payloadData = [NSJSONSerialization dataWithJSONObject:@{kMSPartitions : partitions} options:0 error:&jsonError];
-
-  // Http call.
-  [httpClient sendAsync:payloadData
-      completionHandler:^(NSString *callId, NSHTTPURLResponse *response, NSData *data, NSError *error) {
-        MSLogVerbose([MSDataStore logTag], @"Get token callback, request Id %@ with status code: %lu", callId,
-                     (unsigned long)response.statusCode);
-
-        // If comletion is provided.
-        if (completion) {
-
-          // Read tokens.
-          NSError *tokenResponsejsonError;
-          NSDictionary *jsonDictionary = [NSJSONSerialization JSONObjectWithData:data options:0 error:&tokenResponsejsonError];
-          if (tokenResponsejsonError) {
-            MSLogError([MSDataStore logTag], @"Can't deserialize tokens with error: %@", [tokenResponsejsonError description]);
-            completion([[MSTokensResponse alloc] initWithTokens:nil], error);
-          }
-
-          // Create token result object.
-          MSTokensResponse *tokens = [[MSTokensResponse alloc] initWithDictionary:jsonDictionary];
-          completion(tokens, error);
++ (MSTokenResult *)retrieveCachedToken:(NSString *)partitionName {
+    NSString *tokenString = [MSKeychainUtil stringForKey:[MSTokenExchange tokenKeyNameForPartition:partitionName]];
+    if(tokenString != nil){
+        MSTokenResult *tokenResult = [[MSTokenResult alloc] initWithString:tokenString];
+        
+        if(tokenResult != nil){
+            
+            NSDate *currentUTCDate = [NSDate date];
+            
+            NSDateFormatter* utcFormatter = [[NSDateFormatter alloc] init];
+            [utcFormatter setDateFormat:@"yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"];
```

Comment:
We have a utility method for this format. Please check `MSUtility` class in `MSUtility+Date.m`.